src/consensus/consensus.h

```diff
@@ -12,7 +12,13 @@
 static const unsigned int MAX_BLOCK_SERIALIZED_SIZE = 4000000;
 /** The maximum allowed weight for a block, see BIP 141 (network rule) */
 static const unsigned int MAX_BLOCK_WEIGHT = 4000000;
-/** The maximum allowed size for a block excluding witness data, in bytes (network rule) */
+/**
+ * The maximum allowed size for a block excluding witness data, in bytes (network rule).
+ * This parameter is largely superfluous because it is directly implied by the above block
+ *  weight limit, even when BIP 141 is not active.  It continues to exist for use in
+ *  various early tests that run before the witness data has been checked.
```

Comment:
There are many early checks in validation which limit the resources used handling invalid things or avoid incorrect invalidation in the presence of malleability. Most of them explain themselves.

E.g. in this case you cannot check the weight limit until after you've verified the witness commitment, which takes a bunch of hashing-- otherwise you might invalidate a block you received with an tampered witness (take a valid block add another 2mb of junk to one of the witnesses) instead of just dropping it and the party that gave it to you.  But you can check the base size before doing that.